Append ThinkingSphinx callbacks to Package and Project models

dmarcoux opened the pull request:
Deleted packages/projects are now removed from Sphinx indices. Before, we were missing the `after_destroy` callback from
`ThinkingSphinx::ActiveRecord::Callbacks::DeleteCallbacks`. This resulted in having deleted packages/projects in our Sphinx indices. The error was `ThinkingSphinx::Search::StaleIdsException` with the message `Record IDs found by Sphinx but not by ActiveRecord : 123, 456, 789`.

See upstream:
https://github.com/pat/thinking-sphinx/blob/df2746b5679c8009527ac87b761357d536d48c05/lib/thinking_sphinx/callbacks.rb#L10
https://github.com/pat/thinking-sphinx/blob/7622e5e40fe8a79a3329ddc78d54390cfe7a7f02/lib/thinking_sphinx/callbacks/appender.rb#L16

This fixes #9436.

**TODO after merging this PR:**
We will have to cleanup the indices and to do so, we need to rebuild them.
```shell
# We need to remove the existing indices
cd to_our_application
rm tmp/binlog db/sphinx
run_in_api bundle exec rake ts:rt:rebuild 
```

dmarcoux commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Find solution for TS indexing issues](https://trello.com/c/zALynfru/1448-find-solution-for-ts-indexing-issues)

hennevogel commented on src/api/spec/models/project_spec.rb:
```diff
@@ -768,4 +768,11 @@ def reset_project_in_backend
     it { expect(project.project_state).not_to be_nil }
   end
+
+  describe 'sphinx' do
```
We have `RSpec.describe FullTextSearch`...

dmarcoux replied:
I was looking for exactly that :slightly_smiling_face:

dmarcoux commented:
The specs I added are also passing on master, so without my changes. I really don't get it since I've experienced the opposite in my manual testing. This is what I've done:

1. I created a package in the web UI.
2. I checked if the package is in the indices. It was.
3. I deleted the package in the web UI.
4. I checked if the package is in the indices. It was not.

Without the changes from this PR, the package was still in the indices in the 4th step.

vpereira commented on src/api/app/models/package.rb:
```diff
@@ -21,6 +20,15 @@ class Package < ApplicationRecord
+  ThinkingSphinx::Callbacks.append(self, behaviours: [:real_time]) do |package|
+    # Index package only if its name, title or description changed
+    if package.name_previously_changed? || package.title_previously_changed? || package.description_previously_changed?
+      [package]
```
can you move those callbacks to a concern? like https://api.rubyonrails.org/classes/ActiveSupport/Concern.html ?

dmarcoux replied:
Sure!
